File: src/courses/Classes.java
package courses;


import java.awt.*;

public class Classes{

/* class
A class with fields and method
*/
void point_class() {
class Point {
    int x, y;
    void move(int a, int b) {
        x +=a;
        y+= b;
    }
}
Point p = new Point();
p.move(3, 3);
}

/* inheritance
A Pixel is a Point with color
*/
void inheritance() {
class Pixel extends Point {
    Color color = Color.black;
}
Pixel p = new Pixel();
p.move(3, 3);
p.color = Color.red;
}


/* constructor
A constructor in a init function
*/
void constructor( ){
class Point {
    int x, y;
    Point(int x, int y) {
        this.x = x;
        this.y = y;
    }
}
Point p = new Point(2,2);
}

/* getters and setters

*/
void getters_and_setters() {
class Point {
    int x, y;
    public int getX()       { return x; }
    public void setX(int x) { this.x = x; }
    public int getY()       { return y; }
    public void setY(int y) { this.y = y; }
}
}


/* access modifiers
public, protected, package, private
*/
void access() {
class Point {
    private int a;
    protected int b;
    int c;
    public int d;
}
}

/* override
an Oval is a Shape
*/
void override_example() {
class Shape {
    void draw() {}
}

class Oval extends Shape {
    @Override
    void draw() {}
}
}

/* abstract class
Shape is abstract
*/
void abstract_example() {
abstract class Shape {
    abstract void draw();
}

class Oval extends Shape {
    @Override
    void draw() {}
}
}

/* interface
implements an interface
*/
void interface_example() {
interface Drawable {
    void draw();
}

class Oval implements Drawable {
    @Override
    public void draw() {}
}
}
}

File: src/courses/Classes.kt
package courses

import java.awt.Color
import java.awt.Point

class Classes {
    fun point_class() {
        class Point {
            var x = 0
            var y = 0
            fun move(a: Int, b: Int) {
                x += a
                y += b
            }
        }

        val p = Point()
        p.move(3, 3)
        /// a class with fields and method
    } //

    fun inheritance() {
        class Pixel : Point() {
            var color = Color.black
        }

        val p = Pixel()
        p.move(3, 3)
        p.color = Color.red
        /// A Pixel is a Point with color
    } //

    fun constructor() {
        class Point(var x: Int, var y: Int)

        val p = Point(2, 2)
        /// a constructor in a init function
    } //

    fun getters_and_setters() {
        class Point {
            var x = 0
            var y = 0

        }
        // getters and setters
    } //

    fun access() {
        class Point {
            private val a = 0
            protected var b = 0
            var c = 0
            var d = 0
        }
        /// public, protected, package, private
    } //

    fun override_example() {
        open class Shape {
            open fun draw() {}
        }

        class Oval : Shape() {
            override fun draw() {}
        }
        /// an Oval is a Shape
    } //

    fun abstract_example() {
        abstract class Shape {
            abstract fun draw()
        }

        class Oval : Shape() {
            override fun draw() {}
        }
        /// Shape is abstract
    } //

    internal interface Drawable {
        fun draw()
    }

    fun interface_example() {
        class Oval : Drawable {
            override fun draw() {}
        }
        /// implements an interface
    } //

    companion object {
        fun doe() {}
    }
}
